localstack/services/s3/cors.py

```diff
@@ -0,0 +1,250 @@
+
+# TODO: refactor those to expose the needed methods maybe in another way that both can import
+add_default_headers = CorsResponseEnricher.add_cors_headers
+is_origin_allowed_default = CorsEnforcer.is_cors_origin_allowed
+
+
+class S3CorsHandler(Handler):
+
+    bucket_cors_index: BucketCorsIndex
+
+    def __init__(self):
+        self.bucket_cors_index = BucketCorsIndex()
+        self._service = get_service_catalog().get("s3")
+        self._s3_op_router = RestServiceOperationRouter(self._service)
+
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        self.handle_cors(chain, context, response)
+
+    def pre_parse_s3_request(self, context: RequestContext) -> Tuple[bool, Optional[str]]:
+        is_s3: bool
+        bucket_name: str
+
+        path = context.request.path
+        host = context.request.host
+
+        # first, try to figure out best-effort whether the request is an s3 request
+        if host.startswith(S3_VIRTUAL_HOSTNAME):
+            is_s3 = True
+            bucket_name = path.split("/")[1]
+        # try to extract the bucket from the hostname (the "in" check is a minor optimization
+        elif ".s3" in host and (match := _s3_virtual_host_regex.match(host)):
+            is_s3 = True
+            bucket_name = match.group(3)
+        # otherwise we're not sure, and whether it's s3 depends on whether the bucket exists. check later
+        else:
+            is_s3 = False
+            bucket_name = path.split("/")[1]
+
+        existing_buckets = self.bucket_cors_index.buckets
+        if bucket_name not in existing_buckets:
+            return is_s3, None
+
+        return True, bucket_name
+
+    def handle_cors(self, chain: HandlerChain, context: RequestContext, response: Response):
+        """
+        Handle CORS for S3 requests. S3 CORS rules can be configured.
+        https://docs.aws.amazon.com/AmazonS3/latest/userguide/cors.html
+        https://docs.aws.amazon.com/AmazonS3/latest/userguide/ManageCorsUsing.html
+        """
+
+        # this is used with the new ASF S3 provider
+        # although, we could use it to pre-parse the request and set the context to move the service name parser
+        if config.LEGACY_S3_PROVIDER or config.DISABLE_CUSTOM_CORS_S3:
+            return
+
+        is_s3, bucket_name = self.pre_parse_s3_request(context)
+
+        if not is_s3 and not bucket_name:
+            # continue the chain, let the default CORS handler take care of the request
+            return
+
+        # set the service so that the regular CORS enforcer knows it needs to ignore this request
+        context.service = self._service
+
+        request = context.request
+        is_options_request = request.method == "OPTIONS"
+
+        def stop_options_chain():
+            # FIXME: need to add it there as not handled by the serializer, we stop the chain to avoid the request
+            #  being parsed
```

Comment:
Unfortunately, I don't really understand this comment. Where is there?
If you don't want to trigger the response handlers, you can also call `terminate` on the chain.

Reply:
The `request-id` and such are added by the serializer in the last step of the request handlers. As we stop the chain, those are not added, so we need to add them manually. It was not "there" but "here", sorry. I'm not sure it's a `FIXME`, just a limitation of those fields being added in the serializer.